crates/node_binding/src/dependencies/mod.rs

```diff
@@ -0,0 +1,56 @@
+mod entry_dependency;
+
+use std::{any::TypeId, ffi::c_void, ptr};
+
+pub use entry_dependency::*;
+use napi::{
+  bindgen_prelude::{FromNapiValue, JavaScriptClassExt, Object, ValidateNapiValue},
+  sys, Env, Error, JsObject, NapiRaw, Status,
+};
+use rspack_core::DependencyId;
+use rspack_napi::napi::check_status;
+
+use crate::Dependency;
+
+pub struct DependencyObject(Option<DependencyId>);
+
+impl DependencyObject {
+  pub fn dependency_id(&self) -> Option<DependencyId> {
+    self.0
+  }
+}
+
+impl FromNapiValue for DependencyObject {
+  unsafe fn from_napi_value(
+    env: napi::sys::napi_env,
+    napi_val: napi::sys::napi_value,
+  ) -> napi::Result<Self> {
+    JsObject::validate(env, napi_val)?;
+
+    let js_object = Object::from_napi_value(env, napi_val)?;
+
+    let env_wrapper = Env::from_raw(env);
+    Dependency::instance_of(env_wrapper, &js_object)?;
+
+    let mut unknown_tagged_object: *mut c_void = ptr::null_mut();
+    check_status!(sys::napi_unwrap(
+      env,
+      js_object.raw(),
+      &mut unknown_tagged_object,
+    ))?;
+
+    let type_id = unknown_tagged_object as *const TypeId;
+    if *type_id == TypeId::of::<Dependency>() {
+      let tagged_object = &*(unknown_tagged_object as *mut Dependency);
+      Ok(DependencyObject(Some(tagged_object.dependency_id)))
+    } else if *type_id == TypeId::of::<EntryDependency>() {
+      let tagged_object = &*(unknown_tagged_object as *mut EntryDependency);
+      Ok(DependencyObject(tagged_object.dependency_id))
+    } else {
```

Comment:
I'm not sure whether we can ensure safety here

Reply:
since we already know this has performance cost, maybe we shouldn't include this optimization in this pr